src/param/param_slash.c

```diff
@@ -30,6 +30,85 @@
 static char queue_buf[PARAM_SERVER_MTU];
 param_queue_t param_queue = { .buffer = queue_buf, .buffer_size = PARAM_SERVER_MTU, .type = PARAM_QUEUE_TYPE_EMPTY, .version = 2 };
 
+enum {
+	MISSING_CLOSE_BRACKET = -3,
+	PARAM_NOT_FOUND = -2,
+	EMPTY_ARRAY_SLICE = -1,
+};
+
+static int param_slash_parse_array(char * arg, int node, param_t **param, int *start_index, int *end_index, int *slice_detected) {
+	/**
+	 * Function to find offsets and check if slice delimitor is active.
+	 * @return Returns an array of three values. Defaults to INT_MIN if no value.
+	 * @return Last value is 1 if found.
+	*/
+
+	/* Search for the '@' symbol:
+	 * Call strtok twice in order to skip the stuff head of '@' */
+	
+	char _slice_delimitor;
+	char * saveptr;
+	char * token;
+	
+	int first_scan = 0;
+	int second_scan = 0;
+
+	/* Search for the '[' symbol: */
+	strtok_r(arg, "[", &saveptr);
+	token = strtok_r(NULL, "[", &saveptr);
+	if (token != NULL) {
+		// Check if close bracket exists as last element in token.
+		// If not, then return an error.
+		if(token[strlen(token)-1] != ']'){
+			return MISSING_CLOSE_BRACKET;
+		}
+		// Searches for the format [digit:digit] and [digit:].
+		first_scan = sscanf(token, "%d%c%d", start_index, &_slice_delimitor, end_index);
+		
+		if(first_scan <= 0){
+			// If the input was ":4" then no offsets will be set by the first sscanf, 
+			// so we check for this and try again with another format to match.
+			// second scan will never be 0, since a digit can still be loaded into %c.
+			second_scan = sscanf(token, "%c%d", &_slice_delimitor, end_index);
+		}
+		
+		if(_slice_delimitor == ':'){
+			*slice_detected = 1;
+		}
+		
+		if(second_scan > 0 && _slice_delimitor == ']'){
+			// This is an error, example: set test_array[] 4
+			return EMPTY_ARRAY_SLICE;
+		}
+		
+
+		*token = '\0';
+	}
+
+	char *endptr;
+	int id = strtoul(arg, &endptr, 10);
+	// If strtoul has an error, then it will return ULONG_MAX, so we check on that.
+	if (id != ULONG_MAX && *endptr == '\0') {
+		*param = param_list_find_id(node, id);
+	} else {
+		*param = param_list_find_name(node, arg);
+	}
+
+	if(*param == NULL){
+		return PARAM_NOT_FOUND;
+	}
+
```

Comment:
I would move this part of the code out of this function and leave the responsibility of checking if the parsed parameter array is valid to the caller.
This would allow very easy unit testing of this function ;)

Reply:
Maybe it would be better if param_slash_parse_array() was split into 3 functions: 1 for business logic, 1 for error checking, and a public API that calls both? That way the public API wouldn't have to change (I realize param_slash_parse_array() hasn't been merged yet, but the error check is similar in param_slash_parse()).